Avoid the call to BaseTypeSeq.toList in type Unification.

diesalbla opened the pull request:
The method `toList` of the `BaseTypeSeq` class creates a list from the elements already in an array. In the Types unification, this method was called to create a list of types that was immediately filtered out and folded with an `exists`.

We avoid the call to `toList`, and instead perform a single iteration through the underlying array.

diesalbla committed:
Avoid the call to BaseTypeSeq.toList in type Unification.

The method `toList` of the `BaseTypeSeq` class creates a list
from the elements already in an array. In the Types unification,
this method was called to create a list of types that was
immediately filtered out and folded with an `exists`.

To avoid the call to `toList`, we do the following:
- Add a `toIterator` method in the `BaseTypeSeq` class. This is a
  simple counter that goes through the array at once.
- use the iterator.exists method instead.

Jasper-M commented:
A compromise might be to define a `toIterator` method in `BaseTypeSeq`.

diesalbla commented:
@Jasper-M Indeed, having a `toIterator` method in the `BaseTypeSeq` would be preferable to using a `.toList`. However, using that method would still require one allocation per call. Still better than `n` allocations of `toList`, but the submitted PR has no allocations.

retronym commented:
Let's add `toSeq` to `BaseTyoeSeq` that wraps `elems` in a `immutable.ArraySeq` (this only exists on the 2.13.x branch).

I believe the single allocation cost is totally fine in this context.

retronym reviewed (dismissed):
Let's retarget to 2.12.x and introduce `toSeq`.

diesalbla commented:
Following @Jasper-M comments, I have added a `toIterator` method in the `BaseTypeSeq`. This rather simple iterator is no more than a counter running through the `elems` array. Does anybody know if such a simple iterator is already defined anywhere in the standard library? 

I have changed the code of the unification, to use the `Iterator.exists` method instead of the while loop originally suggested.

diesalbla commented on src/reflect/scala/reflect/internal/BaseTypeSeqs.scala:
```diff
@@ -130,6 +131,17 @@ trait BaseTypeSeqs {
     /** Return all evaluated types in this sequence as a list */
     def toList: List[Type] = elems.toList
 
+    def toIterator: Iterator[Type] = new AbstractIterator[Type] {
```
_Off main topic_: Is this kind of `Iterator` over an `Array` already in the standard collection libary? I could not find it anywhere, and it may come handy for others.

Jasper-M replied:
In 2.13 it exists as the private class `ArrayOps.ArrayIterator`.

retronym replied:
I've pushed a simplification to just use `elems.iterator`

diesalbla replied:
What does the `elems.iterator` do? Where is I defined? Is it by any chance creating a linked sequence and then iterating over it?

retronym replied:
```
scala> (as: Array[String]) => as.iterator //print<TAB>

((as: scala.Array[scala.Predef.String]) => scala.Predef.refArrayOps[String](as).iterator) // : Array[String] => Iterator[String]
```

It implicit converts `Array` to `ArrayOps`, then calls `ArrayOps.iterator` which is basically what you implemented.

diesalbla commented:
> Let's retarget to 2.12.x and introduce toSeq.

@retronym Do you still want to re-target to the `2.13.x` branch, or is the change good enough for both, or would you want to mark this one as _nomerge_ and resubmit another PR with the newer API to `2.13.x`?

retronym commented:
The `toIterator` solution is good enough, let's just stick with that on both branches.
